userspace/libsinsp/metrics_collector.cpp

```diff
@@ -0,0 +1,478 @@
+	if (sum > 0)
+	{
+		cpu_usage_perc_total_host = 100.0 - ((idle * 100.0) / sum);
+		cpu_usage_perc_total_host = std::round(cpu_usage_perc_total_host * 10.0) / 10.0; // round to 1 decimal
+	}
+}
+
+uint64_t metrics_collector::get_container_memory_usage() const
+{
+	/* In Kubernetes `container_memory_working_set_bytes` is the memory measure the OOM killer uses
+	 * and values from `/sys/fs/cgroup/memory/memory.usage_in_bytes` are close enough.
+	 *
+	 * Please note that `kubectl top pod` numbers would reflect the sum of containers in a pod and
+	 * typically libs clients (e.g. Falco) pods contain sidekick containers that use memory as well.
+	 * This metric accounts only for the container with the security monitoring agent running.
+	*/
+	uint64_t memory_used = 0;
+	const char* filepath = getenv(SINSP_AGENT_CGROUP_MEM_PATH_ENV_VAR);
+	if (filepath == nullptr)
+	{
+		filepath = "/sys/fs/cgroup/memory/memory.usage_in_bytes";
+	}
+
+	FILE* f = fopen(filepath, "r");
+	if(!f)
+	{
+		return 0;
+	}
+
+	/* memory size returned in bytes */
+	int fscanf_matched = fscanf(f, "%" SCNu64, &memory_used);
+	fclose(f);
+
+	if (fscanf_matched != 1) {
+		return 0;
+	}
+	return memory_used;
+}
+
+metrics_collector::metrics_collector(sinsp* inspector, const uint32_t& flags, const bool& convert_memory_to_mb) :
+	m_inspector(inspector),
+	m_metrics_flags((METRICS_V2_KERNEL_COUNTERS | METRICS_V2_LIBBPF_STATS | METRICS_V2_RESOURCE_UTILIZATION | METRICS_V2_STATE_COUNTERS)),
+	m_convert_memory_to_mb(true)
+{
+	m_metrics_flags = flags;
+	m_convert_memory_to_mb = convert_memory_to_mb;
+}
+
+metrics_collector::~metrics_collector()
+{
+}
+
+void metrics_collector::snapshot()
+{
+	m_metrics.clear();
+	if (!m_inspector)
+	{
+		return;
+	}
+
+	/* 
+	 * libscap metrics 
+	 */
+
+	if((m_metrics_flags & METRICS_V2_KERNEL_COUNTERS) || (m_metrics_flags & METRICS_V2_LIBBPF_STATS))
+	{
+		uint32_t nstats = 0;
+		int32_t rc = 0;
+		// libscap metrics: m_metrics_flags are pushed down from consumers' input, 
+		// libbpf stats only collected when ENGINE_FLAG_BPF_STATS_ENABLED aka `kernel.bpf_stats_enabled = 1`
+		const metrics_v2* metrics_v2_snapshot = m_inspector->get_capture_stats_v2(m_metrics_flags, &nstats, &rc);
+		if (metrics_v2_snapshot && nstats > 0 && rc == 0)
+		{
+			// Move into m_metrics via std::move
+			m_metrics = std::vector<metrics_v2>(metrics_v2_snapshot, metrics_v2_snapshot + nstats);
+		}
+	}
+
+	/* 
+	 * libsinsp metrics 
+	 */
+
+	if((m_metrics_flags & METRICS_V2_RESOURCE_UTILIZATION))
+	{
+		const scap_agent_info* agent_info = m_inspector->get_agent_info();
+		uint32_t rss{0}, vsz{0}, pss{0}, procs_running_host{0};
+		uint64_t memory_used_host{0}, open_fds_host{0};
+		double cpu_usage_perc{0.0}, cpu_usage_perc_total_host{0.0};
+		uint64_t container_memory_usage = get_container_memory_usage();
+		metrics_v2_value_unit rss_unit{METRIC_VALUE_UNIT_MEMORY_KILOBYTES}, vsz_unit{METRIC_VALUE_UNIT_MEMORY_KILOBYTES}, \
```

Comment:
What is the backslash at the end-of-line for? Line continuation? It's not needed in C++, unless we talk about the preprocessor.